packages/slice-machine/src/features/autoSave/useAutoSave.tsx

```diff
@@ -0,0 +1,114 @@
+import { useCallback, useState, useEffect } from "react";
+import { uniqueId } from "lodash";
+import { toast } from "react-toastify";
+import { Box, Button, Text } from "@prismicio/editor-ui";
+
+type UseAutoSaveArgs = {
+  debounceDelay?: number;
+  errorMessage?: string;
+  retryDelay?: number;
+  retryMessage?: string;
+};
+
+type AutoSaveState = {
+  pendingSave: (() => Promise<unknown>) | null;
+  nextSave: (() => Promise<unknown>) | null;
+};
+
+export type AutoSaveStatusType = "saving" | "saved" | "error";
+
+export const useAutoSave = (args: UseAutoSaveArgs = {}) => {
+  const {
+    debounceDelay = 500,
+    errorMessage = "An error happened while saving",
+    retryDelay = 2000,
+    retryMessage = "Retry",
+  } = args;
+  const [autoSaveState, setAutoSaveState] = useState<AutoSaveState>({
+    pendingSave: null,
+    nextSave: null,
+  });
+  const [autoSaveStatusInternal, setAutoSaveStatusInternal] =
+    useState<AutoSaveStatusType>("saved");
+  const [autoSaveStatus, setAutoSaveStatus] = useState<AutoSaveStatusType>(
+    autoSaveStatusInternal,
+  );
+
+  const setNextSave = useCallback((saveAction: () => Promise<unknown>) => {
+    setAutoSaveState((prevState) => ({
+      ...prevState,
+      nextSave: saveAction,
+    }));
+  }, []);
+
+  const executeSave = useCallback(async () => {
+    if (autoSaveState.pendingSave) {
+      setAutoSaveStatusInternal("saving");
+
+      try {
+        await autoSaveState.pendingSave();
+
+        setAutoSaveStatusInternal("saved");
+        setAutoSaveState((prevState) => ({
+          pendingSave: prevState.nextSave,
+          nextSave: null,
+        }));
+      } catch (error) {
+        setAutoSaveStatusInternal("error");
+        console.error(errorMessage, error);
+
+        const toastId = uniqueId();
+        toast.error(
+          <Box gap={8} alignItems="center">
+            <Text>{errorMessage}</Text>
+            <Button
+              onClick={() => {
+                toast.dismiss(toastId);
+                setAutoSaveStatusInternal("saving");
+                setTimeout(() => {
+                  void executeSave();
+                }, retryDelay);
+              }}
+            >
+              {retryMessage}
+            </Button>
+          </Box>,
+          {
+            autoClose: false,
+            closeOnClick: false,
+            draggable: false,
+            toastId,
+          },
+        );
+      }
+    }
+  }, [autoSaveState, errorMessage, retryDelay, retryMessage]);
+
+  useEffect(() => {
+    if (
+      autoSaveStatusInternal !== "saving" &&
+      (autoSaveState.pendingSave || autoSaveState.nextSave)
+    ) {
+      setAutoSaveState((prevState) => ({
+        pendingSave: prevState.nextSave,
+        nextSave: null,
+      }));
+
+      executeSave().catch((error) =>
+        console.error("Error in save effect:", error),
+      );
+    }
+  }, [autoSaveStatusInternal, autoSaveState, executeSave]);
```

Comment:
I believe every property of `autoSaveState` should be listed in here

Reply:
Hey, it's not necessary here since it's immutable, we are talking about a state so each time the values changed the state itself 100% change. You never modify a state sub value directly.

The linter is pretty good on that and tell us here that it's good.